cpp/include/kvikio/stream.hpp

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <sys/types.h>
+#include <algorithm>
+#include <cstdlib>
+#include <kvikio/error.hpp>
+#include <kvikio/shim/cuda.hpp>
+#include <kvikio/shim/cufile.hpp>
+#include <tuple>
+
+namespace kvikio {
+
+/**
+ * @brief Future of an asynchronous IO operation
+ *
+ * This class shouldn't be used directly, instead some stream operations such as
+ * `FileHandle.read_async` and `FileHandle.write_async` returns an instance of this class. Use
+ * `.check_bytes_done()` to synchronize the associated CUDA stream and return the number of bytes
+ * read or written by the operation.
+ *
+ * The goal of this class is twofold:
+ *   - Have `read_async` and `write_async` return an object that clearly associates the function
+ *     arguments with the CUDA stream used. This is useful because the current validity of the
+ *     arguments depends on the stream.
+ *   - Support of by-value arguments. In many cases, a user will use `read_async` and `write_async`
+ *     like most other asynchronous CUDA functions that take by-value arguments.
+ *
+ * To support by-value arguments, we allocate the arguments on the heap (malloc `ArgByVal`) and have
+ * the by-reference arguments (`ArgByRef`) points into `ArgByVal`. This way, the `read_async` and
+ * `write_async` can call `.get_args()` to get the by-reference arguments required by cuFile's
+ * stream API. However, this also means that the caller of `read_async` and `write_async` MUST keep
+ * the returned `StreamFuture` alive until the operations is done otherwise `StreamFuture` will free
+ * `ArgByVal` before cuFile had a change to use them!
+ */
+class StreamFuture {
+ public:
+  struct ArgByVal {
+    std::size_t size;
+    off_t file_offset;
+    off_t devPtr_offset;
+    ssize_t bytes_done;
+  };
+  struct ArgByRef {
+    std::size_t* size_p;
+    off_t* file_offset_p;
+    off_t* devPtr_offset_p;
+    ssize_t* bytes_done_p;
+  };
+
+ private:
+  void* _devPtr_base{nullptr};
+  CUstream _stream{nullptr};
+  ArgByVal* _val{nullptr};
+  ArgByRef _ref{nullptr};
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏Do we actually need both of these stored in the class, does it suffice instead to do something like:

```c++
class StreamFuture {
  private:
     struct data {
        std::size_t size;
        off_t file_offset;
        off_t devptr_offset;
        ssize_t bytes_done;
    }
    data* _values{};
  public:
    StreamFuture(...args) 
    {
        _values = static_cast<data *>(...);
        _values->size = size;
        _values->...;
     }
     std::tuple<...> get_args() { 
        return {&_values->size, &_values->file_offset, ...};
    }
}
```
Or is there some subtlety I am missing?